packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js

```diff
@@ -0,0 +1,365 @@
+        arr,
+        prevNode,
+        heightsArr,
+        itemsSoFar = [],
+      }: {|
+        item: T,
+        i: number,
+        arr: $ReadOnlyArray<T>,
+        prevNode: NodeData<T>,
+        heightsArr: $ReadOnlyArray<number>,
+        itemsSoFar?: $ReadOnlyArray<T>,
+      |}) {
+        // Copy the heights array so we don't mutate
+        const heightsSoFar = [...heightsArr];
+
+        // Get the positions and heights after adding this item
+        const { positions: updatedPositions, heights: updatedHeights } =
+          getOneColumnItemPositions<T>({
+            items: [...itemsSoFar, item],
+            heights: heightsSoFar,
+            ...commonGetPositionArgs,
+          });
+
+        // Add the new node to the graph
+        const paintedItemData = {
+          id: item,
+          heights: updatedHeights,
+          positions: updatedPositions,
+        };
+
+        const adjacentColumnHeightDeltas = getAdjacentColumnHeightDeltas(updatedHeights);
+        const lowestAdjacentColumnHeightDelta = Math.min(...adjacentColumnHeightDeltas);
+
+        graph.addNode(paintedItemData);
+        graph.addEdge(prevNode, paintedItemData, lowestAdjacentColumnHeightDelta);
+
+        // If there are items remaining in the array that haven't yet been laid out, keep going
+        if (arr.length > 1) {
+          const otherItems = [...arr];
+          otherItems.splice(i, 1);
+          otherItems.forEach((otherItem, index, array) => {
+            addPossibleLayout({
+              item: otherItem,
+              i: index,
+              arr: array,
+              heightsArr,
+              prevNode: paintedItemData,
+              itemsSoFar: [...itemsSoFar, item],
+            });
+          });
+        }
+      }
+
+      // For each unpainted item, start generating possible layouts
+      oneColumnItems.forEach((item, i, arr) => {
+        addPossibleLayout({
+          item,
+          i,
+          arr,
+          heightsArr: paintedItemHeights,
+          prevNode: startNodeData,
+        });
+      });
+
+      const { lowestScore, lowestScoreNode } = graph.findLowestScore(startNodeData);
+
+      // The best solution may be "no solution", i.e. laying out the 2-col item first
+      const winningNode =
+        lowestScore === null || lowestScore < startingLowestAdjacentColumnHeightDelta
+          ? lowestScoreNode
+          : startNodeData;
+
+      const { positions: winningPositions } = winningNode;
+
+      // Insert 2-col item(s)
+      const twoColItem = twoColumnItems[0]; // this should always only be one
+      const { heights: finalHeights, position: twoColItemPosition } = getTwoColItemPosition<T>({
+        item: twoColItem,
+        heights: lowestScoreNode.heights,
+        ...commonGetPositionArgs,
+      });
+
+      // Combine winning positions and 2-col item position, add to cache
+      const finalPositions = [
+        ...winningPositions,
+        { item: twoColItem, position: twoColItemPosition },
```

Comment:
thinking ahead, once we integrate this into Pinboard it would be good to somehow have access to this so we can log data on average number of whitespace we end up with per pass.  maybe something like a callback prop could help with that e.g. `onTwoColumnLayout(position, heights)`

Reply:
Yeah I was thinking about that as well. I think I'll add that in a future PR so we can start playing around with these changes in Pinboard